### PR TITLE
[sample_vpp] Change mfxVideoParams to smart wrappers

### PR DESCRIPTION
**Pull request by AntonGrishin**

### CODE REVIEW

File: samples/sample_vpp/src/sample_vpp_utils.cpp [hunk 1]
```diff
@@ -780,12 +780,14 @@ void WipeMemoryAllocator(sMemoryAllocator* pAllocator)
 
 void WipeConfigParam( sAppResources* pResources )
 {
-
-    if( pResources->multiViewConfig.View )
+    auto multiViewConfig = pResources->pVppParams->GetExtBuffer<mfxExtMVCSeqDesc>();
+    if (multiViewConfig)
     {
-        delete [] pResources->multiViewConfig.View;
+        if (multiViewConfig->View)
```
Comment (mgonchar) [inline samples/sample_vpp/src/sample_vpp_utils.cpp]: remove this ```if``` deleting ```nullptr``` is a valid operation which simply does nothing

File: samples/sample_vpp/src/sample_vpp_utils.cpp [hunk 2]
```diff
@@ -817,10 +819,14 @@ void WipeResources(sAppResources* pResources)
         pResources->pDstFileWriters = NULL;
     }
 
-    if(pResources->compositeConfig.InputStream)
+    auto compositeConfig = pResources->pVppParams->GetExtBuffer<mfxExtVPPComposite>();
+    if(compositeConfig)
     {
-        delete[] pResources->compositeConfig.InputStream;
-        pResources->compositeConfig.InputStream=NULL;
+        if(compositeConfig->InputStream)
```
Comment (mgonchar) [inline samples/sample_vpp/src/sample_vpp_utils.cpp]: same

File: samples/sample_vpp/src/sample_vpp_config.cpp [hunk 3]
```diff
@@ -243,72 +202,57 @@ mfxStatus ConfigVideoEnhancementFilters( sInputParams* pParams, sAppResources* p
...
     if( VPP_FILTER_ENABLED_CONFIGURED == pParams->compositionParam.mode )
     {
-        pResources->compositeConfig.Header.BufferId = MFX_EXTBUFF_VPP_COMPOSITE;
-        pResources->compositeConfig.Header.BufferSz = sizeof(mfxExtVPPComposite);
-        pResources->compositeConfig.NumInputStream  = pParams->numStreams;
-        pResources->compositeConfig.InputStream     = new mfxVPPCompInputStream[pResources->compositeConfig.NumInputStream];
-        memset(pResources->compositeConfig.InputStream, 0, sizeof(mfxVPPCompInputStream) * pResources->compositeConfig.NumInputStream);
+        auto compositeConfig             = pVppParam->AddExtBuffer<mfxExtVPPComposite>();
+        compositeConfig->NumInputStream  = pParams->numStreams;
+        compositeConfig->InputStream     = new mfxVPPCompInputStream[compositeConfig->NumInputStream];
```
Comment (mgonchar) [inline samples/sample_vpp/src/sample_vpp_config.cpp]: ```compositeConfig->InputStream     = new mfxVPPCompInputStream[compositeConfig->NumInputStream]();```
and delete explicit zeroing

File: samples/sample_vpp/src/sample_vpp_utils.cpp (hunk 2 above)
Comment (AntonGrishin) [reply samples/sample_vpp/src/sample_vpp_utils.cpp]: done

File: samples/sample_vpp/src/sample_vpp_utils.cpp (hunk 1 above)
Comment (AntonGrishin) [reply samples/sample_vpp/src/sample_vpp_utils.cpp]: done

File: samples/sample_vpp/src/sample_vpp_config.cpp (hunk 3 above)
Comment (AntonGrishin) [reply samples/sample_vpp/src/sample_vpp_config.cpp]: Done

**Comment by AntonGrishin** [comment]
> Seems mirroring got broken

Fixed

File: samples/sample_common/include/sample_utils.h [hunk 4]
```diff
@@ -257,6 +253,33 @@ template<>struct mfx_ext_buffer_id<mfxExtHEVCRegion> {
...
+    enum {id = MFX_EXTBUFF_VPP_ROTATION};
+};
+template<>struct mfx_ext_buffer_id<mfxExtVPPScaling> {
+    enum {id = MFX_EXTBUFF_VPP_SCALING};
+};
+template<>struct mfx_ext_buffer_id<mfxExtColorConversion> {
+    enum {id = MFX_EXTBUFF_VPP_COLOR_CONVERSION};
+};
 
 constexpr uint16_t max_num_ext_buffers = 38 * 2; // '*2' is for max estimation if all extBuffer were 'paired'
```
Comment (dmitryermilov) [inline samples/sample_common/include/sample_utils.h]: please update

File: samples/sample_vpp/src/sample_vpp_config.cpp [hunk 5]
```diff
@@ -77,122 +67,94 @@ mfxStatus ConfigVideoEnhancementFilters( sInputParams* pParams, sAppResources* p
 
     if (enhFilterCount > 0)
     {
-        pResources->extDoUse.NumAlg = enhFilterCount;
-        pResources->extDoUse.AlgList = pResources->tabDoUseAlg;
-        pVppParam->ExtParam[pVppParam->NumExtParam++] = (mfxExtBuffer*)&(pResources->extDoUse);
+        auto doUse = pVppParam->AddExtBuffer<mfxExtVPPDoUse>();
+        doUse->NumAlg = enhFilterCount;
+        doUse->AlgList = pResources->tabDoUseAlg;
```
Comment (dvrogozh) [inline samples/sample_vpp/src/sample_vpp_config.cpp]: why we need `tabDoUseAlg` with the smart adapter?

File: samples/sample_vpp/include/sample_vpp_utils.h [hunk 6]
```diff
@@ -462,47 +462,21 @@ struct sAppResources
...
-    mfxExtVPPComposite     compositeConfig;
-
+    sFrameProcessor*       pProcessor;
+    MfxVideoParamsWrapper* pVppParams;
+    sMemoryAllocator*      pAllocator;
+    sInputParams*          pParams;
+    SurfaceVPPStore*       pSurfStore;
+
+    // number of video enhancement filters (denoise, procamp, detail, video_analysis, multi_view, ste, istab, tcc, ace, svc)
+    constexpr static uint32_t ENH_FILTERS_COUNT = 20;
```
Comment (dvrogozh) [inline samples/sample_vpp/include/sample_vpp_utils.h]: this hardcoded value makes we wonder when we will crash

File: samples/sample_vpp/src/sample_vpp_config.cpp (hunk 5 above)
Comment (AntonGrishin) [reply samples/sample_vpp/src/sample_vpp_config.cpp]: I try to safe logic, which we used before (you can check comment [1])

File: samples/sample_vpp/include/sample_vpp_utils.h (hunk 6 above)
Comment (AntonGrishin) [reply samples/sample_vpp/include/sample_vpp_utils.h]: moved from
https://github.com/Intel-Media-SDK/MediaSDK/blob/b8ac0a6bb6c8d32f08066b765dda27bd8a09be17/samples/sample_vpp/include/sample_vpp_config.h#L46

File: samples/sample_common/include/sample_utils.h (hunk 4 above)
Comment (AntonGrishin) [reply samples/sample_common/include/sample_utils.h]: done

File: samples/sample_vpp/src/sample_vpp_config.cpp (hunk 5 above)
Comment (dmitryermilov) [reply samples/sample_vpp/src/sample_vpp_config.cpp]: @dvrogozh , please see lines 32-65. AlgList  is a pointer, we need to initialize it with some memory. For mfxExtBuffer with underlying memory, we allocate this memory outside of extBufferHolder<T> class.

**Pull request merged**